SerialPrograms/Source/CommonFramework/Inference/SpectrogramMatcher.cpp

```diff
@@ -0,0 +1,203 @@
+#include "SpectrogramMatcher.h"
+#include "CommonFramework/AudioPipeline/AudioTemplate.h"
+#include "CommonFramework/Tools/AudioFeed.h"
+
+#include <iostream>
+#include <cfloat>
+
+namespace PokemonAutomation{
+
+
+SpectrogramMatcher::SpectrogramMatcher(const QString& templateFilename, Mode mode) : m_mode(mode) {
+    m_template = loadAudioTemplate(templateFilename);
+    const size_t numTemplateWindows = m_template.numWindows();
+    m_numOriginalFrequencies = m_template.numFrequencies();
+    if (m_template.numFrequencies() == 0){  // Error case, failed to load template
+        return;
+    }
+
+    // TODO: get sample rate as part of the AudioFeed interface:
+    const int sampleRate = 48000;
+    const int halfSampleRate = sampleRate / 2;
+
+     // The frquency range from [0.0, (numTemplateFrequencies-1)/numTemplateFrequencies * halfSampleRate]
+    // Since human can only hear as high as 20KHz sound, matching on frequencies >= 20KHz is meaningless.
+    // So the index i of the max frequency we should be matching is the one with
+    // i /numTemplateFrequencies * halfSampleRate <= 20KHz
+    // i <= numTemplateFrequencies * 20K / halfSampleRate
+
+    m_originalFreqStart = 5;
+    m_originalFreqEnd = 20000 * m_numOriginalFrequencies / halfSampleRate + 1;
+    
+    if (m_mode == Mode::SPIKE_CONV){
+        // Do convolution on audio template
+        const size_t numConvedFrequencies = (m_originalFreqEnd - m_originalFreqStart) - m_spikeKernel.size() + 1;
+        std::vector<float> temporaryBuffer(numConvedFrequencies * numTemplateWindows);
+        for(size_t i = 0; i < numTemplateWindows; i++){
+            conv(m_template.getWindow(i) + m_originalFreqStart, m_originalFreqEnd - m_originalFreqStart,
+                temporaryBuffer.data() + i * numConvedFrequencies);
+        }
+
+        m_template = AudioTemplate(std::move(temporaryBuffer), numTemplateWindows);
+        m_freqStart = 0;
+        m_freqEnd = numConvedFrequencies;
+    } else{
+        m_freqStart = m_originalFreqStart;
+        m_freqEnd = m_originalFreqEnd;
+    }
+    
+    m_templateNorm = templateNorm();
+}
+
+size_t SpectrogramMatcher::latestTimestamp() const{
+    if (m_spectrums.size() == 0){
+        return SIZE_MAX;
+    }
+    return m_spectrums.front()->stamp;
+}
+
+void SpectrogramMatcher::conv(const float* src, size_t num, float* dst){
+    if (num < m_spikeKernel.size()){
+        return;
+    }
+    for(size_t i = 0; i < num-m_spikeKernel.size()+1; i++){

```

Comment:
Reminder for myself to profile this step to make sure it's not consuming an unreasonable amount of CPU.